drivers/network/dd/b57xx/interrupt.c

```diff
@@ -0,0 +1,177 @@
+/*
+ * PROJECT:     ReactOS Broadcom NetXtreme Driver
+ * LICENSE:     GPL-2.0-or-later (https://spdx.org/licenses/GPL-2.0-or-later)
+ * PURPOSE:     Interrupt handlers
+ * COPYRIGHT:   Copyright 2021-2022 Scott Maday <coldasdryice1@gmail.com>
+ */
+
+#include "nic.h"
+
+#define NDEBUG
+#include <debug.h>
+
+VOID
+NTAPI
+MiniportDisableInterrupt(_In_ NDIS_HANDLE MiniportAdapterContext)
+{
+    PB57XX_ADAPTER Adapter = (PB57XX_ADAPTER)MiniportAdapterContext;
+    
+    NICDisableInterrupts(Adapter);
+}
+
+VOID
+NTAPI
+MiniportEnableInterrupt(_In_ NDIS_HANDLE MiniportAdapterContext)
+{
+    PB57XX_ADAPTER Adapter = (PB57XX_ADAPTER)MiniportAdapterContext;
+    
+    NICEnableInterrupts(Adapter);
+}
+
+VOID
+NTAPI
+MiniportISR(_Out_ PBOOLEAN InterruptRecognized,
+            _Out_ PBOOLEAN QueueMiniportHandleInterrupt,
+            _In_  NDIS_HANDLE MiniportAdapterContext)
+{
+    PB57XX_ADAPTER Adapter = (PB57XX_ADAPTER)MiniportAdapterContext;
+    
+    if (Adapter->Status.pBlock->Status & B57XX_SB_UPDATED && Adapter->Interrupt.Pending == FALSE)
+    {
+        NICInterruptAcknowledge(Adapter);
+        
+        Adapter->Status.LastTag = Adapter->Status.pBlock->StatusTag;
+        Adapter->Status.pBlock->Status &= ~B57XX_SB_UPDATED;
+        
+        Adapter->Interrupt.Pending = TRUE;
+        
+        *InterruptRecognized = TRUE;
+        *QueueMiniportHandleInterrupt = TRUE;
+    }
+    else
+    {
+        *InterruptRecognized = FALSE;
+        *QueueMiniportHandleInterrupt = FALSE;
+    }
+}
+
+VOID
+NTAPI
+MiniportHandleInterrupt(_In_ NDIS_HANDLE MiniportAdapterContext)
+{
+    ULONG ProducerIndex;
+    ULONG ConsumerIndex;
+    PCHAR BufferBase;
+    PB57XX_RECEIVE_BUFFER_DESCRIPTOR pReceiveDescriptor;
+    PB57XX_ADAPTER Adapter = (PB57XX_ADAPTER)MiniportAdapterContext;
+    BOOLEAN LinkStateChange = FALSE;
+    BOOLEAN GotAny = FALSE;
+    
+    if (Adapter->HardwareStatus != NdisHardwareStatusReady)
+    {
+        return;
+    }
+    
+    NdisDprAcquireSpinLock(&Adapter->Interrupt.Lock);
+    if (Adapter->Interrupt.Pending == FALSE)
+    {
+        return;
+    }
+    Adapter->Interrupt.Pending = FALSE;
+    
+    do
+    {
+        Adapter->Status.pBlock->Status &= ~B57XX_SB_UPDATED;
+        
+        /* Process any link changes */
+        if (Adapter->Status.pBlock->Status & B57XX_SB_LINKSTATE)
+        {
+            Adapter->Status.pBlock->Status &= ~B57XX_SB_LINKSTATE;
+            
+            LinkStateChange = TRUE;
+            NDIS_MinDbgPrint("Link state change\n");
+            
+            NdisMIndicateStatus(Adapter->MiniportAdapterHandle,
+                                NICUpdateLinkStatus(Adapter),
+                                NULL,
+                                0);
+            NdisMIndicateStatusComplete(Adapter->MiniportAdapterHandle);
```

Comment:
Same for NdisMEthIndicateReceive, NdisMEthIndicateReceiveComplete, NdisMSendComplete...

Reply:
> Same for NdisMEthIndicateReceive, NdisMEthIndicateReceiveComplete, NdisMSendComplete.

Note: drivers/network/dd/rtl8139/interrupt.c does not release the spinlock before calling these